src/librustc_borrowck/indexed_set.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright 2012-2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// FIXME: move this to `rustc_data_structures`
+
+use std::fmt;
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::{Deref, DerefMut, Range};
+use bitslice::{BitSlice, Word};
+use bitslice::{bitwise, Union, Subtract};
+
+pub trait Indexed {
+    type Idx: Idx;
+}
+
+pub trait Idx: 'static {
+    fn new(usize) -> Self;
+    fn idx(&self) -> usize;
+}
+
+pub struct OwnIdxSet<T: Idx> {
+    _pd: PhantomData<fn(&[T], usize) -> &T>,
```

Comment:
Oh this `PhantomData` got things mixed up; you are right that `PhantomData<T>` would suffice here.

(I had originally attempted to put the actual *represented* type into the interface, e.g. make a connection from `MovePathIndex` to `MovePath`, and then the goal was to have a phantom data that would say `Fn(&[MovePath], MovePathIndex) -> Option<&MovePath>`. But as I reworked the API and removed references to the referenced data type, I didn't notice that I had mixed up the roles of `T` (the index type) and the referenced data type (now unnamed here).

I will replace with `PhantomData<Fn(&T)>` since that more properly reflects the relationship between the `IdxSet` and the `Idx` that each is fed.